File: src/main/java/org/compiere/orm/MOrg.java
package org.compiere.orm;

import kotliquery.Row;
import org.compiere.model.I_AD_Org;
import org.idempiere.common.util.CCache;

import java.util.List;
import java.util.Properties;

import static software.hsharp.core.util.DBKt.getSQLValue;

/**
 * Organization Model
 *
 * @author Jorg Janke
 * @version $Id: MOrg.java,v 1.3 2006/07/30 00:58:04 jjanke Exp $
 */
public class MOrg extends X_AD_Org {
    /**
     *
     */
    private static final long serialVersionUID = -5604686137606338725L;
    /**
     * Cache
     */
    private static CCache<Integer, MOrg> s_cache = new CCache<Integer, MOrg>(I_AD_Org.Table_Name, 50);
    /**
     * Linked Business Partner
     */
    private Integer m_linkedBPartner = null;

    /**
     * ************************************************************************ Standard Constructor
     *
     * @param ctx       context
     * @param AD_Org_ID id
     * @param trxName   transaction
     */
    public MOrg(Properties ctx, int AD_Org_ID) {
        super(ctx, AD_Org_ID);
        if (AD_Org_ID == 0) {
            //	setValue (null);
            //	setName (null);
            setIsSummary(false);
        }
    } //	MOrg

    /**
     * Load Constructor
     *
     * @param ctx     context
     * @param rs      result set
     * @param trxName transaction
     */
    public MOrg(Properties ctx, Row row) {
        super(ctx, row);
    } //	MOrg

    /**
     * Parent Constructor
     *
     * @param client client
     * @param name   name
     */
    public MOrg(MClient client, String value, String name) {
        this(client.getCtx(), 0);
        setADClientID(client.getClientId());
        setSearchKey(value);
        setName(name);
    } //	MOrg

    /**
     * Get Active Organizations Of Client
     *
     * @param po persistent object
     * @return array of orgs
     */
    public static MOrg[] getOfClient(PO po) {
        List<MOrg> list =
                new Query(po.getCtx(), I_AD_Org.Table_Name, "AD_Client_ID=?")
                        .setOrderBy(I_AD_Org.COLUMNNAME_Value)
                        .setOnlyActiveRecords(true)
                        .setParameters(po.getClientId())
                        .list();
        for (MOrg org : list) {
            s_cache.put(org.getId(), org);
        }
        return list.toArray(new MOrg[list.size()]);
    } //	getOfClient

    /**
     * Get Org from Cache
     *
     * @param ctx       context
     * @param AD_Org_ID id
     * @return MOrg
     */
    public static MOrg get(Properties ctx, int AD_Org_ID) {
        MOrg retValue = s_cache.get(AD_Org_ID);
        if (retValue != null) return retValue;
        retValue = new MOrg(ctx, AD_Org_ID);
        if (retValue.getId() != 0) s_cache.put(AD_Org_ID, retValue);
        return retValue;
    } //	get

    /**
     * Get Org Info
     *
     * @return Org Info
     */
    public MOrgInfo getInfo() {
        return MOrgInfo.get(getCtx(), getOrgId());
    } //	getMOrgInfo

    /**
     * After Save
     *
     * @param newRecord new Record
     * @param success   save success
     * @return success
     */
    protected boolean afterSave(boolean newRecord, boolean success) {
        if (!success) return success;
        if (newRecord) {
            //	Info
            MOrgInfo info = new MOrgInfo(this);
            info.saveEx();
            //	Access
            MRoleOrgAccess.createForOrg(this);
            MRole role = MRole.getDefault(getCtx(), true); // 	reload
            role.loadAccess(true); // reload org access within transaction
            //	TreeNode
            insert_Tree(MTree_Base.TREETYPE_Organization);
        }
        if (newRecord || is_ValueChanged(I_AD_Org.COLUMNNAME_Value))
            update_Tree(MTree_Base.TREETYPE_Organization);

        return true;
    } //	afterSave

    /**
     * After Delete
     *
     * @param success
     * @return deleted
     */
    protected boolean afterDelete(boolean success) {
        if (success) delete_Tree(MTree_Base.TREETYPE_Organization);
        return success;
    } //	afterDelete

    /**
     * Get Linked BPartner
     *
     * @return C_BPartner_ID
     */
    public int getLinkedC_BPartner_ID(String trxName) {
        if (m_linkedBPartner == null) {
            int C_BPartner_ID =
                    getSQLValue(
                            "SELECT C_BPartner_ID FROM C_BPartner WHERE AD_OrgBP_ID=?", getOrgId());
            if (C_BPartner_ID < 0) // 	not found = -1
                C_BPartner_ID = 0;
            m_linkedBPartner = new Integer(C_BPartner_ID);
        }
        return m_linkedBPartner.intValue();
    } //	getLinkedC_BPartner_ID
} //	MOrg

File: src/main/kotlin/org/compiere/orm/MOrg.kt
package org.compiere.orm

import kotliquery.Row
import org.compiere.model.ClientOrganization
import org.idempiere.common.util.factory
import org.idempiere.common.util.loadUsing
import org.idempiere.common.util.memoize
import software.hsharp.core.util.asResource

import software.hsharp.core.util.getSQLValue

private val orgFactory = factory { MOrg(it) }

/**
 * Get organization by Id
 */
fun getOrg(orgId: Int) = orgId loadUsing orgFactory

/**
 * Get Active Organizations Of Client
 *
 * @param po persistent object
 * @return array of orgs
 */
fun getClientOrganizations(po: PO): Array<ClientOrganization> {
    val list = Query<ClientOrganization>(ClientOrganization.Table_Name, "AD_Client_ID=?")
        .setOrderBy(ClientOrganization.COLUMNNAME_Value)
        .setOnlyActiveRecords(true)
        .setParameters(po.clientId)
        .list()
    return list.map { getOrg(it.id) }.toTypedArray()
} // 	getOfClient

/**
 * Organization Model
 *
 * @author Jorg Janke
 * @version $Id: MOrg.java,v 1.3 2006/07/30 00:58:04 jjanke Exp $
 */
class MOrg : X_AD_Org {
    private val doLinkedBusinessPartnerId = {
        val result = "/sql/getLinkedBusinessPartnerId.sql".asResource { sql ->
            getSQLValue(sql, orgId)
        }
        if (result < 0) 0 else result
    }.memoize()

    /**
     * Get Org Info
     *
     * @return Org Info
     */
    val info: MOrgInfo?
        get() = getOrganizationInfo(orgId) // 	getMOrgInfo

    /**
     * Get Linked BPartner
     *
     * @return C_BPartner_ID
     */
    // 	not found = -1
    val linkedBusinessPartnerId: Int
        get() = doLinkedBusinessPartnerId()

    /**
     * ************************************************************************ Standard Constructor
     *
     * @param AD_Org_ID id
     */
    constructor(AD_Org_ID: Int) : super(AD_Org_ID) {
        if (AD_Org_ID == 0) {
            setIsSummary(false)
        }
    } // 	MOrg

    /**
     * Load Constructor
     *
     */
    constructor(row: Row) : super(row) // 	MOrg

    /**
     * Parent Constructor
     *
     * @param client client
     * @param orgName name
     */
    constructor(client: MClient, value: String, orgName: String) : this(0) {
        setClientId(client.clientId)
        searchKey = value
        name = orgName
    } // 	MOrg

    /**
     * After Save
     *
     * @param newRecord new Record
     * @param success save success
     * @return success
     */
    override fun afterSave(newRecord: Boolean, success: Boolean): Boolean {
        if (!success) return success
        if (newRecord) {
            // 	Info
            val info = MOrgInfo(this)
            info.saveEx()
            // 	Access
            MRoleOrgAccess.createForOrg(this)
            val role = getDefaultRole(true) // 	reload
            role.loadAccess(true) // reload org access within transaction
            // 	TreeNode
            insertTree(MTree_Base.TREETYPE_Organization)
        }
        if (newRecord || isValueChanged(ClientOrganization.COLUMNNAME_Value))
            updateTree(MTree_Base.TREETYPE_Organization)

        return true
    } // 	afterSave

    /**
     * After Delete
     *
     * @param success
     * @return deleted
     */
    override fun afterDelete(success: Boolean): Boolean {
        if (success) deleteTree(MTree_Base.TREETYPE_Organization)
        return success
    } // 	afterDelete

    companion object {
        private const val serialVersionUID = -5604686137606338725L
    }
} // 	MOrg
